File: src/main/java/jdi/demo/SimpleDebugger.java
package jdi.demo;

import com.sun.jdi.*;
import com.sun.jdi.connect.Connector;
import com.sun.jdi.connect.IllegalConnectorArgumentsException;
import com.sun.jdi.connect.LaunchingConnector;
import com.sun.jdi.connect.VMStartException;
import com.sun.jdi.event.*;
import com.sun.jdi.request.BreakpointRequest;
import com.sun.jdi.request.ClassPrepareRequest;
import com.sun.jdi.request.EventRequest;
import com.sun.jdi.request.EventRequestManager;

import java.io.IOException;
import java.util.List;
import java.util.Map;


class SimpleDebugger {

    private VirtualMachine vm;
    private EventRequestManager eventRequestManager;
    private EventSet eventSet;
    private boolean vmExit = false;

    void launch() {

        log("start for debugger");
        vm = launchTarget("jdi.demo.Hello");

        Process process = vm.process();

        // Register ClassPrepareRequest
        eventRequestManager = vm.eventRequestManager();
        ClassPrepareRequest classPrepareRequest
                = eventRequestManager.createClassPrepareRequest();
        classPrepareRequest.addClassFilter("jdi.demo.Hello");
        classPrepareRequest.addCountFilter(1);
        classPrepareRequest.setSuspendPolicy(EventRequest.SUSPEND_ALL);
        classPrepareRequest.enable();

        // Enter event loop
        try {
            eventLoop();
        } catch (Exception e) {
            e.printStackTrace();
        }

        process.destroy();
    }

    /**
     * 参数：
     * mainArgs为目标程序main函数所在的类
     **/
    private VirtualMachine launchTarget(String mainArgs) {
        //findLaunchingConnector：获取连接
        LaunchingConnector connector = findLaunchingConnector();
        //connectorArguments：设置连接参数
        Map arguments = connectorArguments(connector, mainArgs);
        try {
            return connector.launch(arguments);//启动连接
        } catch (IOException exc) {
            throw new Error("Unable to launch target VM: " + exc);
        } catch (IllegalConnectorArgumentsException exc) {
            throw new Error("Internal error: " + exc);
        } catch (VMStartException exc) {
            throw new Error("Target VM failed to initialize: " + exc.getMessage());
        }
    }

    private LaunchingConnector findLaunchingConnector() {
        VirtualMachineManager vmm = Bootstrap.virtualMachineManager();
        return vmm.defaultConnector();
    }

    /**
     * 参数：
     * connector为清单1.中获取的Connector连接实例
     * mainArgs为目标程序main函数所在的类
     **/
    private Map connectorArguments(LaunchingConnector connector, String mainArgs) {
        Map arguments = connector.defaultArguments();
        Connector.Argument mainArg = (Connector.Argument) arguments.get("main");
        if (mainArg == null) {
            throw new Error("Bad launching connector");
        }
        mainArg.setValue(mainArgs);
        return arguments;
    }

    private void eventLoop() throws Exception {
        EventQueue eventQueue = vm.eventQueue();
        while (!vmExit) {
            eventSet = eventQueue.remove();
            EventIterator eventIterator = eventSet.eventIterator();
            while (eventIterator.hasNext()) {
                Event event = eventIterator.next();
                execute(event);
            }
        }
    }

    private void execute(Event event) throws Exception {
        if (event instanceof VMStartEvent) {
            log("----> VMStartEvent");
            eventSet.resume();
        } else if (event instanceof ClassPrepareEvent) {
            log("----> ClassPrepareEvent");
            ClassPrepareEvent classPrepareEvent = (ClassPrepareEvent) event;
            String mainClassName = classPrepareEvent.referenceType().name();
            if (mainClassName.equals("jdi.demo.Hello")) {
                log("Class " + mainClassName + " is already prepared");
            }
            // Get location
            ReferenceType referenceType = classPrepareEvent.referenceType();
            List locations = referenceType.locationsOfLine(10);
            Location location = (Location) locations.get(0);

            // Create BreakpointEvent
            BreakpointRequest breakpointRequest = eventRequestManager
                    .createBreakpointRequest(location);
            breakpointRequest.setSuspendPolicy(EventRequest.SUSPEND_ALL);
            breakpointRequest.enable();
            eventSet.resume();
        } else if (event instanceof BreakpointEvent) {
            log("----> BreakpointEvent");
            log("Reach line 10 of jdi.demo.Hello");
            BreakpointEvent breakpointEvent = (BreakpointEvent) event;
            ThreadReference threadReference = breakpointEvent.thread();
            StackFrame stackFrame = threadReference.frame(0);
            LocalVariable localVariable = stackFrame
                    .visibleVariableByName("str");
            Value value = stackFrame.getValue(localVariable);
            String str = ((StringReference) value).value();
            log("The local variable str at line 10 is " + str + " of " + value.type().name());
            eventSet.resume();
        } else if (event instanceof VMDisconnectEvent) {
            vmExit = true;
            log("----> VMDisconnectEvent");
        } else {
            log("----> " + event.toString());
            eventSet.resume();
        }
    }

    private static void log(String content) {
        System.out.println(content);
    }

}

File: src/main/kotlin/jdi/client/SimpleDebugger.kt
package jdi.client

import com.sun.jdi.Bootstrap
import com.sun.jdi.Location
import com.sun.jdi.StringReference
import com.sun.jdi.VirtualMachine
import com.sun.jdi.event.*
import com.sun.jdi.request.EventRequest
import com.sun.jdi.request.EventRequestManager


class SimpleDebugger {

    private lateinit var vm: VirtualMachine
    private lateinit var process: Process
    private lateinit var eventRequestManager: EventRequestManager
    private lateinit var eventQueue: EventQueue
    private lateinit var eventSet: EventSet
    private var vmExit = false

    fun init() {
        val launchingConnector = Bootstrap.virtualMachineManager().defaultConnector()

        // Get arguments of the launching connector
        val defaultArguments = launchingConnector.defaultArguments()
        val mainArg = defaultArguments["main"]
        val suspendArg = defaultArguments["suspend"]
        // Set class of main method
        mainArg!!.setValue("jdi.demo.Hello")
        suspendArg!!.setValue("true")
        vm = launchingConnector.launch(defaultArguments)

        process = vm.process()

        // Register ClassPrepareRequest
        eventRequestManager = vm.eventRequestManager()
        val classPrepareRequest = eventRequestManager.createClassPrepareRequest()
        classPrepareRequest.addClassFilter("jdi.demo.Hello")
        classPrepareRequest.addCountFilter(1)
        classPrepareRequest.setSuspendPolicy(EventRequest.SUSPEND_ALL)
        classPrepareRequest.enable()

        // Enter event loop
        eventLoop()

        process.destroy()
    }

    @Throws(Exception::class)
    private fun eventLoop() {
        eventQueue = vm.eventQueue()
        while (true) {
            if (vmExit) {
                break
            }
            eventSet = eventQueue.remove()
            val eventIterator = eventSet.eventIterator()
            while (eventIterator.hasNext()) {
                val event = eventIterator.next() as Event
                execute(event)
            }
        }
    }

    @Throws(Exception::class)
    private fun execute(event: Event) {
        if (event is VMStartEvent) {
            println("VM started")
            eventSet.resume()
        } else if (event is ClassPrepareEvent) {
            val mainClassName = event.referenceType().name()
            if (mainClassName == "jdi.demo.Hello") {
                println("Class $mainClassName is already prepared")
            }
            // Get location
            val referenceType = event.referenceType()
            val locations = referenceType.locationsOfLine(10)
            val location = locations.get(0) as Location

            // Create BreakpointEvent
            val breakpointRequest = eventRequestManager
                .createBreakpointRequest(location)
            breakpointRequest.setSuspendPolicy(EventRequest.SUSPEND_ALL)
            breakpointRequest.enable()
            eventSet.resume()
        } else if (event is BreakpointEvent) {
            println("Reach line 10 of jdi.demo.Hello")
            val threadReference = event.thread()
            val stackFrame = threadReference.frame(0)
            val localVariable = stackFrame
                .visibleVariableByName("str")
            val value = stackFrame.getValue(localVariable)
            val str = (value as StringReference).value()
            println(
                ("The local variable str at line 10 is $str of ${value.type().name()}")
            )
            eventSet.resume()
        } else if (event is VMDisconnectEvent) {
            vmExit = true
        } else {
            eventSet.resume()
        }
    }

}

fun main() {
    val debugger = SimpleDebugger()
    debugger.init()

}
